linux/free.1
.Dd June 30, 2023
.Dt FREE 1
.Os
.Sh NAME
.Nm free
.Nd Display amount of free and used memory in the system

.Sh SYNOPSIS
.Nm
.Op options

.Sh DESCRIPTION
free displays the total amount of free and used physical and swap memory in the system, as well as the buffers and caches used by the kernel. The information is gathered by parsing /proc/meminfo. The displayed columns are:

.Bl -tag -width indent
.It total
Total installed memory (MemTotal and SwapTotal in /proc/meminfo).
.It used
Used memory (calculated as total - free - buffers - cache).
.It free
Unused memory (MemFree and SwapFree in /proc/meminfo).
.It shared
Memory used (mostly) by tmpfs (Shmem in /proc/meminfo).
.It buffers
Memory used by kernel buffers (Buffers in /proc/meminfo).
.It cache
Memory used by the page cache and slabs (Cached and SReclaimable in /proc/meminfo). 
.It buff/cache
Sum of buffers and cache.
.It available
Estimation of how much memory is available for starting new applications, without swapping. Unlike the data provided by the cache or free fields, this field takes into account page cache and also that not all reclaimable memory slabs will be reclaimed due to items being in use (MemAvailable in /proc/meminfo, available on kernels 3.14, emulated on kernels 2.6.27+, otherwise the same as free).
.El

.Sh OPTIONS
.Bl -tag -width indent
.It Fl b, -bytes
Display the amount of memory in bytes.
.It Fl k, -kibi
Display the amount of memory in kibibytes. This is the default.
.It Fl m, -mebi
Display the amount of memory in mebibytes.
.It Fl g, -gibi
Display the amount of memory in gibibytes.
.It Fl t, -tebi
Display the amount of memory in tebibytes.
.It Fl p, -pebi
Display the amount of memory in pebibytes.
.It Fl e, -exbi
Display the amount of memory in exbibytes.
.It Fl -kilo
Display the amount of memory in kilobytes. Implies --si.
.It Fl -mega
Display the amount of memory in megabytes. Implies --si.
.It Fl -giga
Display the amount of memory in gigabytes. Implies --si.
.It Fl -tera
Display the amount of memory in terabytes. Implies --si.
.It Fl -peta
Display the amount of memory in petabytes. Implies --si.
.It Fl -exa
Display the amount of memory in exabytes. Implies --si.
.It Fl -yaml
Output the data as YAML.
.It Fl -json
Output the data as JSON.
.It Fl w, -wide
Switch to the wide mode. The wide mode produces lines longer than 80 characters. In this mode buffers and cache are reported in two separate columns.
.It Fl -si
Use kilo, mega, giga etc (power of 1000) instead of kibi, mebi, gibi (power of 1024).
.It Fl l, -lohi
Show detailed low and high memory statistics.
.It Fl s, -seconds
Continuously display the result N seconds apart.
.It Fl c, -count
Display the result count times. Requires the -s option.
.It Fl t, -total
Display a line showing the column totals.
.It Fl h, -help
Print help.
.It Fl V, -version
Display version information.
.El

.Sh FILES
.Bl -tag -width indent
.It /proc/meminfo
memory information.
.El

.Sh REPORTING BUGS
You can report bugs to <gdanko@gmail.com>.

.Sh AUTHOR
This port was written by Gary Danko.

.Sh SEE ALSO
ps(1), slabtop(1), top(1), vmstat(8)


free.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"runtime"

	"strconv"
	"strings"
	"time"

	"github.com/jcelliott/lumber"
	flags "github.com/jessevdk/go-flags"

	"github.com/shirou/gopsutil/v3/mem"
	"gopkg.in/yaml.v3"
)

const VERSION = "0.3.2"

var (
	logger *lumber.ConsoleLogger
)

type Options struct {
	Bytes   bool   `short:"b" long:"bytes" description:"Show output in bytes."`
	Kilo    bool   `long:"kilo" description:"Show output in kilobytes. Implies --si."`
	Mega    bool   `long:"mega" description:"Show output in megabytes. Implies --si."`
	Giga    bool   `long:"giga" description:"Show output in gigabytes. Implies --si."`
	Tera    bool   `long:"tera" description:"Show output in terabytes. Implies --si."`
	Peta    bool   `long:"peta" description:"Show output in petabytes. Implies --si."`
	Exa     bool   `long:"exa" description:"Show output in exabytes. Implies --si."`
	Kibi    bool   `short:"k" long:"kibi" description:"Show output in kibibytes."`
	Mebi    bool   `short:"m" long:"mebi" description:"Show output in mebibytes."`
	Gibi    bool   `short:"g" long:"gibi" description:"Show output in gibibytes."`
	Tebi    bool   `short:"t" long:"tebi" description:"Show output in tebibytes."`
	Pebi    bool   `short:"p" long:"pebi" description:"Show output in pebibytes."`
	Exbi    bool   `short:"e" long:"exbi" description:"Show output in exbibytes."`
	Json    bool   `short:"j" long:"json" description:"Output the data as JSON."`
	Yaml    bool   `short:"y" long:"yaml" description:"Output the data as YAML."`
	Si      bool   `long:"si" description:"Use kilo, mega, giga, etc (power of 1000) instead of kibi, mebi, gibi (power of 1024)."`
	LoHi    bool   `short:"l" long:"lohi" description:"Show detailed low and high memory statistics (Linux Only)."`
	Total   bool   `long:"total" description:"Show total for RAM + swap."`
	Seconds int    `short:"s" long:"seconds" description:"Continuously display the result N seconds apart."`
	Count   int    `short:"c" long:"count" description:"Display the result N times."`
	Wide    bool   `short:"w" long:"wide" description:"Wide output (Linux Only)."`
	Version func() `short:"V" long:"version" description:"Output version information and exit."`
}

type Memory struct {
	Active    uint64 `json:"active" yaml:"active"`
	Available uint64 `json:"available" yaml:"available"`
	Buffers   uint64 `json:"buffers" yaml:"buffers"`
	Cached    uint64 `json:"cached" yaml:"cached"`
	Free      uint64 `json:"free" yaml:"free"`
	Inactive  uint64 `json:"inactive" yaml:"inactive"`
	Shared    uint64 `json:"shared" yaml:"shared"`
	Total     uint64 `json:"total" yaml:"total"`
	Used      uint64 `json:"used" yaml:"used"`
	Wired     uint64 `json:"wired" yaml:"wired"`
}

type HighMemory struct {
	Total uint64 `json:"total" yaml:"total"`
	Used  uint64 `json:"used" yaml:"used"`
	Free  uint64 `json:"free" yaml:"free"`
}

type LowMemory struct {
	Total uint64 `json:"total" yaml:"total"`
	Used  uint64 `json:"used" yaml:"used"`
	Free  uint64 `json:"free" yaml:"free"`
}

type Swap struct {
	Total uint64 `json:"total" yaml:"total"`
	Used  uint64 `json:"used" yaml:"used"`
	Free  uint64 `json:"free" yaml:"free"`
}

type Totals struct {
	Total uint64 `json:"total" yaml:"total"`
	Used  uint64 `json:"used" yaml:"used"`
	Free  uint64 `json:"free" yaml:"free"`
}

func getBaseExponent(opts Options) (base, exponent float64, abbreviation, prefix string) {
	var (
		defaultAbbreviation string  = "KiB"
		defaultBase         float64 = 1024
		defaultExponent     float64 = 1
		defaultPrefix       string  = "kibi"
	)

	if opts.Bytes {
		base = 1024
		exponent = 0
		abbreviation = "B"
		prefix = ""
	} else if opts.Kibi {
		base = 1024
		exponent = 1
		abbreviation = "KiB"
		prefix = "kibi"
	} else if opts.Mebi {
		base = 1024
		exponent = 2
		abbreviation = "MiB"
		prefix = "mebi"
	} else if opts.Gibi {
		base = 1024
		exponent = 3
		abbreviation = "GiB"
		prefix = "gibi"
	} else if opts.Tebi {
		base = 1024
		exponent = 4
		abbreviation = "TiB"
		prefix = "tebi"
	} else if opts.Pebi {
		base = 1024
		exponent = 5
		abbreviation = "PiB"
		prefix = "pebi"
	} else if opts.Exbi {
		base = 1024
		exponent = 6
		abbreviation = "EiB"
		prefix = "exbi"
	} else if opts.Kilo {
		base = 1000
		exponent = 1
		abbreviation = "KB"
		prefix = "kilo"
	} else if opts.Mega {
		base = 1000
		exponent = 2
		abbreviation = "MB"
		prefix = "mega"
	} else if opts.Giga {
		base = 1000
		exponent = 3
		abbreviation = "GB"
		prefix = "giga"
	} else if opts.Tera {
		base = 1000
		exponent = 4
		abbreviation = "TB"
		prefix = "tera"
	} else if opts.Peta {
		base = 1000
		exponent = 5
		abbreviation = "PB"
		prefix = "peta"
	} else if opts.Exa {
		base = 1000
		exponent = 5
		abbreviation = "EB"
		prefix = "exa"

	} else {
		abbreviation = defaultAbbreviation
		base = defaultBase
		exponent = defaultExponent
		prefix = defaultPrefix
	}

	if opts.Si {
		base = 1000
	}

	return base, exponent, abbreviation, prefix
}

func gatherData() (memoryStats Memory, swapStats Swap, lowMemoryStats LowMemory, highMemoryStats HighMemory, totalStats Totals, err error) {
	memoryInfo, _ := mem.VirtualMemory()
	memoryStats = Memory{
		Active:    memoryInfo.Active,
		Available: memoryInfo.Available,
		Buffers:   memoryInfo.Buffers,
		Cached:    memoryInfo.Cached,
		Free:      memoryInfo.Free,
		Inactive:  memoryInfo.Inactive,
		Shared:    memoryInfo.Shared,
		Total:     memoryInfo.Total,
		Used:      memoryInfo.Used,
		Wired:     memoryInfo.Wired,
	}

	swapStats = Swap{
		Total: memoryInfo.SwapTotal,
		Free:  memoryInfo.SwapFree,
		Used:  memoryInfo.SwapTotal - memoryInfo.SwapFree,
	}

	lowMemoryStats = LowMemory{
		Total: memoryInfo.LowTotal,
		Free:  memoryInfo.LowFree,
		Used:  memoryInfo.LowTotal - memoryInfo.LowFree,
	}

	highMemoryStats = HighMemory{
		Total: memoryInfo.HighTotal,
		Free:  memoryInfo.HighFree,
		Used:  memoryInfo.HighTotal - memoryInfo.HighFree,
	}

	totalStats = Totals{
		Total: memoryStats.Total + swapStats.Total,
		Free:  memoryStats.Free + swapStats.Free,
		Used:  memoryStats.Used + swapStats.Used,
	}

	return memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats, err
}

func generateJSON(opts Options, memoryOut, swapOut, totalsOut map[string]string, descriptor string) (jsonText string, err error) {
	jsonOut := make(map[string]interface{})
	jsonOut["descriptor"] = descriptor
	jsonOut["memory"] = memoryOut
	jsonOut["swap"] = swapOut
	if opts.Total {
		jsonOut["total"] = totalsOut
	}

	jsonBytes, _ := json.Marshal(jsonOut)
	var prettyJSON bytes.Buffer
	if err = json.Indent(&prettyJSON, jsonBytes, "", "    "); err != nil {
		return "", err

	}

	return prettyJSON.String(), err
}

func generateYAML(opts Options, memoryOut, swapOut, totalsOut map[string]string, descriptor string) (jsonText string, err error) {
	yamlOut := make(map[string]interface{})
	yamlOut["descriptor"] = descriptor
	yamlOut["memory"] = memoryOut
	yamlOut["swap"] = swapOut
	if opts.Total {
		yamlOut["total"] = totalsOut
	}

	yamlBytes, _ := yaml.Marshal(yamlOut)

	yamlString := string(yamlBytes)
	yamlString = strings.TrimSpace(yamlString)
	return yamlString, err
}

func displayOutput(opts Options, base, exponent float64, prefix string, memoryStats Memory, swapStats Swap, lowMemoryStats LowMemory, highMemoryStats HighMemory, totalStats Totals) {
	descriptor := prefix + "bytes"
	divideBy := uint64(math.Pow(base, exponent))

	highMemoryFree := highMemoryStats.Free / divideBy
	highMemoryTotal := highMemoryStats.Total / divideBy
	highMemoryUsed := highMemoryStats.Used / divideBy
	lowMemoryFree := highMemoryStats.Free / divideBy
	lowMemoryTotal := highMemoryStats.Total / divideBy
	lowMemoryUsed := highMemoryStats.Used / divideBy
	memActive := memoryStats.Active / divideBy
	memAvailable := memoryStats.Available / divideBy
	memBuffers := memoryStats.Buffers / divideBy
	memCached := memoryStats.Cached / divideBy
	memFree := memoryStats.Free / divideBy
	memInactive := memoryStats.Inactive / divideBy
	memShared := memoryStats.Shared / divideBy
	memTotal := memoryStats.Total / divideBy
	memUsed := memoryStats.Used / divideBy
	memWired := memoryStats.Wired / divideBy
	swapFree := swapStats.Free / divideBy
	swapTotal := swapStats.Total / divideBy
	swapUsed := swapStats.Used / divideBy

	memoryOut := map[string]string{
		"available": strconv.Itoa(int(memAvailable)),
		"free":      strconv.Itoa(int(memFree)),
		"used":      strconv.Itoa(int(memUsed)),
		"total":     strconv.Itoa(int(memTotal)),
	}
	if runtime.GOOS == "darwin" {
		memoryOut["active"] = strconv.Itoa(int(memActive))
		memoryOut["inactive"] = strconv.Itoa(int(memInactive))
		memoryOut["wired"] = strconv.Itoa(int(memWired))
	} else if runtime.GOOS == "linux" {
		memoryOut["buffers"] = strconv.Itoa(int(memBuffers))
		memoryOut["cache"] = strconv.Itoa(int(memCached))
		memoryOut["shared"] = strconv.Itoa(int(memShared))
	}

	swapOut := map[string]string{
		"total": strconv.Itoa(int(swapTotal)),
		"used":  strconv.Itoa(int(swapUsed)),
		"free":  strconv.Itoa(int(swapFree)),
	}

	lowMemoryOut := map[string]string{
		"total": strconv.Itoa(int(lowMemoryTotal)),
		"used":  strconv.Itoa(int(lowMemoryUsed)),
		"free":  strconv.Itoa(int(lowMemoryFree)),
	}

	highMemoryOut := map[string]string{
		"total": strconv.Itoa(int(highMemoryTotal)),
		"used":  strconv.Itoa(int(highMemoryUsed)),
		"free":  strconv.Itoa(int(highMemoryFree)),
	}

	totalsOut := map[string]string{
		"total": strconv.Itoa(int(memTotal) + int(swapTotal)),
		"used":  strconv.Itoa(int(memUsed) + int(swapUsed)),
		"free":  strconv.Itoa(int(memFree) + int(swapFree)),
	}

	if opts.Json {
		jsonText, err := generateJSON(opts, memoryOut, swapOut, totalsOut, descriptor)
		if err != nil {
			logger.Error(err.Error())
			os.Exit(1)
		}
		fmt.Println(jsonText)

	} else if opts.Yaml {
		yamlText, err := generateYAML(opts, memoryOut, swapOut, totalsOut, descriptor)
		if err != nil {
			logger.Error(err.Error())
			os.Exit(1)
		}
		fmt.Println(yamlText)

	} else {
		if runtime.GOOS == "darwin" {
			fmt.Printf("  %18s %11s %11s %11s %11s %11s %11s\n", "total", "used", "free", "active", "inactive", "wired", "available")
			fmt.Printf("Mem:     %11s %11s %11s %11s %11s %11s %11s\n",
				memoryOut["total"],
				memoryOut["used"],
				memoryOut["free"],
				memoryOut["active"],
				memoryOut["inactive"],
				memoryOut["wired"],
				memoryOut["available"],
			)
		} else if runtime.GOOS == "linux" {
			if opts.Wide {
				fmt.Printf("  %18s %11s %11s %11s %11s %11s %11s\n", "total", "used", "free", "shared", "buffers", "cache", "available")
				fmt.Printf("Mem:     %11s %11s %11s %11s %11s %11s %11s\n",
					memoryOut["total"],
					memoryOut["used"],
					memoryOut["free"],
					memoryOut["shared"],
					memoryOut["buffers"],
					memoryOut["cache"],
					memoryOut["available"],
				)
			} else {
				memBuffCache := memBuffers + memCached
				fmt.Printf("  %18s %11s %11s %11s %11s %11s\n", "total", "used", "free", "shared", "buff/cache", "available")
				fmt.Printf("Mem:     %11s %11s %11s %11s %11s %11s\n",
					memoryOut["total"],
					memoryOut["used"],
					memoryOut["free"],
					memoryOut["shared"],
					strconv.Itoa(int(memBuffCache)),
					memoryOut["available"],
				)
			}

			if opts.LoHi {
				fmt.Printf("Low:    %12s %11s %11s\n",
					lowMemoryOut["total"],
					lowMemoryOut["used"],
					lowMemoryOut["free"],
				)
				fmt.Printf("High:    %11s %11s %11s\n",
					highMemoryOut["total"],
					highMemoryOut["used"],
					highMemoryOut["free"],
				)
			}
		}
		fmt.Printf("Swap:    %11s %11s %11s\n",
			swapOut["total"],
			swapOut["used"],
			swapOut["free"],
		)
		if opts.Total {
			fmt.Printf("Total: %13s %11s %11s\n",
				totalsOut["total"],
				totalsOut["used"],
				totalsOut["free"],
			)
		}
	}
}

func main() {
	logger = lumber.NewConsoleLogger(lumber.INFO)

	opts := Options{}

	opts.Version = func() {
		fmt.Printf("free version %s\n", VERSION)
		os.Exit(0)
	}

	parser := flags.NewParser(&opts, flags.Default)
	if _, err := parser.Parse(); err != nil {
		if flagsErr, ok := err.(*flags.Error); ok && flagsErr.Type == flags.ErrHelp {
			os.Exit(0)
		} else {
			logger.Error(err.Error())
			os.Exit(1)
		}
	}

	// Find the base, exponent, and prefix, based on (k, m, g, etc)
	base, exponent, _, prefix := getBaseExponent(opts)

	if opts.Seconds == 0 && opts.Count == 0 {
		memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats, err := gatherData()
		if err != nil {
			logger.Error(err.Error())
			os.Exit(1)
		}
		displayOutput(opts, base, exponent, prefix, memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats)
	} else if opts.Seconds > 0 && opts.Count == 0 {
		for {
			memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats, err := gatherData()
			if err != nil {
				logger.Error(err.Error())
				os.Exit(1)
			}
			displayOutput(opts, base, exponent, prefix, memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats)
			fmt.Println()
			time.Sleep(time.Duration(opts.Seconds) * time.Second)
		}
	} else if opts.Count > 0 {
		if opts.Seconds == 0 {
			logger.Error("--count requires --seconds")
			os.Exit(1)
		}
		for i := 1; i < opts.Count+1; i++ {
			memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats, err := gatherData()
			if err != nil {
				logger.Error(err.Error())
				os.Exit(1)
			}
			displayOutput(opts, base, exponent, prefix, memoryStats, swapStats, lowMemoryStats, highMemoryStats, totalStats)
			if i != opts.Count {
				fmt.Println()
				time.Sleep(time.Duration(opts.Seconds) * time.Second)
			}
		}
	}
}
